Implement the compiling of population node.

tcsnfkx opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/9)
<!-- Reviewable:end -->

tcsnfkx committed:
Implement the compiling of population node.

efoxepstein reviewed (changes requested):
> Reviewed 16 of 16 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/9)* status: all files reviewed, 8 unresolved discussions (waiting on @EvgSkv, @kunyuchen, @lenjoy, @mpatel24, @my2457, @tcsnfkx, and @wliue)

___
*[src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc, line 158 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/9#-Mn88GBXELyrtI7mJH0l:-Mn88GBYAQKS5H9GRn2G:bj9zp6d) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/2c0a17121e2462237bc6f89ddbe4d973c268d265/src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc#L158)):*
> ```cpp
> }
> 
> absl::Status SantizeAdf(const ActivityDensityFunction& adf) {
> ```

Santize? Maybe "`Validate`"?

To me, the difference between sanitizing and validating is this: sanitizing cleans up the input by modifying it or producing a modified copy. Validating just checks whether or not it's acceptable.
___
*[src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc, line 185 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/9#-Mn88bP3AdfGpRq0wEle:-Mn88bP3AdfGpRq0wElf:bob1mmb) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/2c0a17121e2462237bc6f89ddbe4d973c268d265/src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc#L185)):*
> ```cpp
> }
> 
> absl::StatusOr<absl::flat_hash_map<
> ```

This type is a little complex, can you add some classes/structs/typealiases to make this more meaningful?

You could imagine an interface like:

```
class CountryRegionMap {
  const RegionMap& GetRegions(const Country&);
};

class RegionMap {
  const MultipoolRecordSet& GetPool(const Region&);
}
```

or something similar with any additional functionality you need.
___
*[src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc, line 281 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/9#-Mn8CLYYAgeC1mVtUj7F:-Mn8CLYYAgeC1mVtUj7G:bs5rpp1) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/2c0a17121e2462237bc6f89ddbe4d973c268d265/src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc#L281)):*
> ```cpp
>     boundaries[i + 1] = population_sum * alphas[i];
>   }
>   std::partial_sum(boundaries.begin(), boundaries.end(), boundaries.begin());
> ```

More efficient, fewer lines of code, and arguably simpler to reason about (for those less familiar with the standard library):

```
  std::vector<double> boundaries(alphas.size() + 1, 0.0);
  for (int i = 0; i < alphas.size(); ++i) {
    boundaries[i + 1] = population_sum * alphas[i] + boundaries[i];
  }
```
___
*[src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc, line 293 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/9#-Mn8B-6y2odquErLQ3WY:-Mn8B-6y2odquErLQ3WZ:b-4v8q9) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/2c0a17121e2462237bc6f89ddbe4d973c268d265/src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc#L293)):*
> ```cpp
>                            discretized_boundaries.end(), differences.begin());
> 
>   std::vector<uint64_t> output(differences.begin() + 1, differences.end());
> ```

More compact and more efficient:

```
for (int i = 0; i < discretized_boundaries.size() - 1; ++i) {
  discretized_boundaries[i] = discretized_boundaries[i+1] - discretized_boundaries[i];
}
discretized_boundaries.pop_back();
return discretized_boundaries;
```
___
*[src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc, line 368 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/9#-Mn8DMwXEGXv1b_dxaf0:-Mn8DMwXEGXv1b_dxaf1:b-hvmziu) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/2c0a17121e2462237bc6f89ddbe4d973c268d265/src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc#L368)):*
> ```cpp
> 
> // Compiling population pool to a BranchNode.
> absl::StatusOr<BranchNode> CompilePopulationPool(
> ```

This functions seems pretty long, can you extract any helper functions to make it more readable?
___
*[src/main/cc/wfa/virtual_people/training/model_compiler/field_filter_utils.cc, line 61 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/9#-Mn8DghJCQ8ibr4r_5XI:-Mn8DghJCQ8ibr4r_5XJ:binsmd3) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/2c0a17121e2462237bc6f89ddbe4d973c268d265/src/main/cc/wfa/virtual_people/training/model_compiler/field_filter_utils.cc#L61)):*
> ```cpp
> 
>     // Remove any sub_filters which is TRUE filter.
>     filter.mutable_sub_filters()->erase(
> ```

For those less familiar with C++ (like me), maybe include a link in a comment to https://en.wikipedia.org/wiki/Erase%E2%80%93remove_idiom.
___
*[src/test/cc/wfa/virtual_people/training/model_compiler/compiler_test.cc, line 425 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/9#-Mn8ErIOA4DcnIbfVcf6:-Mn8ErIP5C8Wq-Ce9Xkr:brp1ivq) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/2c0a17121e2462237bc6f89ddbe4d973c268d265/src/test/cc/wfa/virtual_people/training/model_compiler/compiler_test.cc#L425)):*
> ```cpp
>                 alphas: 0.6
>                 deltas { coordinates: 1.9 }
>                 deltas { coordinates: 0.4 }
> ```

These textprotos are pretty large. Consider moving them into `.textproto` files? That can be useful because we get syntax highlighting, validation, etc. in IDEs for textprotos.
___
*[src/test/cc/wfa/virtual_people/training/model_compiler/test_data/activity_density_function.textproto, line 5 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/9#-Mn8F3hhD3P-PdC9hckL:-Mn8F3hhD3P-PdC9hckM:bqt1ejo) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/2c0a17121e2462237bc6f89ddbe4d973c268d265/src/test/cc/wfa/virtual_people/training/model_compiler/test_data/activity_density_function.textproto#L5)):*
> ```
>   op: TRUE
> }
> idenitifer_type_names: "TestIdType"
> ```

`identifier` looks like it's spelled wrong


<!-- Sent from Reviewable.io -->

tcsnfkx committed:
Fix based on comments.

Refactor helper function name.
Use typealias to improve the readability.
Remove unnecessary copies in SplitPopulationByAlphas.
Update comments.
Move reading textproto function to common-cpp.
Move textproto strings into files.
Fix typo in ADF proto.

tcsnfkx committed:
Split CompilePopulationPool to improve code readability.

tcsnfkx reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/9)* status: 6 of 22 files reviewed, 8 unresolved discussions (waiting on @efoxepstein, @EvgSkv, @kunyuchen, @lenjoy, @mpatel24, @my2457, and @wliue)

___
*[src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc, line 158 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/9#-Mn88GBXELyrtI7mJH0l:-MnXQIXb2fQ7KGn-dn79:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/2c0a17121e2462237bc6f89ddbe4d973c268d265/src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc#L158)):*
<details><summary><i>Previously, efoxepstein (Eli Fox-Epstein) wrote…</i></summary><blockquote>

Santize? Maybe "`Validate`"?

To me, the difference between sanitizing and validating is this: sanitizing cleans up the input by modifying it or producing a modified copy. Validating just checks whether or not it's acceptable.
</blockquote></details>

Done.
___
*[src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc, line 185 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/9#-Mn88bP3AdfGpRq0wEle:-MnXQP235DKJQAU4mZkv:b-gg8760) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/2c0a17121e2462237bc6f89ddbe4d973c268d265/src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc#L185)):*
<details><summary><i>Previously, efoxepstein (Eli Fox-Epstein) wrote…</i></summary><blockquote>

This type is a little complex, can you add some classes/structs/typealiases to make this more meaningful?

You could imagine an interface like:

```
class CountryRegionMap {
  const RegionMap& GetRegions(const Country&);
};

class RegionMap {
  const MultipoolRecordSet& GetPool(const Region&);
}
```

or something similar with any additional functionality you need.
</blockquote></details>

Use type aliases.
___
*[src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc, line 281 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/9#-Mn8CLYYAgeC1mVtUj7F:-MnXQ_P_66ZzT3poKFXQ:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/2c0a17121e2462237bc6f89ddbe4d973c268d265/src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc#L281)):*
<details><summary><i>Previously, efoxepstein (Eli Fox-Epstein) wrote…</i></summary><blockquote>

More efficient, fewer lines of code, and arguably simpler to reason about (for those less familiar with the standard library):

```
  std::vector<double> boundaries(alphas.size() + 1, 0.0);
  for (int i = 0; i < alphas.size(); ++i) {
    boundaries[i + 1] = population_sum * alphas[i] + boundaries[i];
  }
```
</blockquote></details>

Done.
___
*[src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc, line 293 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/9#-Mn8B-6y2odquErLQ3WY:-MnXQbHp-M9SEJqGmNzA:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/2c0a17121e2462237bc6f89ddbe4d973c268d265/src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc#L293)):*
<details><summary><i>Previously, efoxepstein (Eli Fox-Epstein) wrote…</i></summary><blockquote>

More compact and more efficient:

```
for (int i = 0; i < discretized_boundaries.size() - 1; ++i) {
  discretized_boundaries[i] = discretized_boundaries[i+1] - discretized_boundaries[i];
}
discretized_boundaries.pop_back();
return discretized_boundaries;
```
</blockquote></details>

Done.
___
*[src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc, line 368 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/9#-Mn8DMwXEGXv1b_dxaf0:-MnX_EhY5zbHih4imemn:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/2c0a17121e2462237bc6f89ddbe4d973c268d265/src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc#L368)):*
<details><summary><i>Previously, efoxepstein (Eli Fox-Epstein) wrote…</i></summary><blockquote>

This functions seems pretty long, can you extract any helper functions to make it more readable?
</blockquote></details>

Done.
___
*[src/main/cc/wfa/virtual_people/training/model_compiler/field_filter_utils.cc, line 61 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/9#-Mn8DghJCQ8ibr4r_5XI:-MnXQfmu4-T286M6uAE4:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/2c0a17121e2462237bc6f89ddbe4d973c268d265/src/main/cc/wfa/virtual_people/training/model_compiler/field_filter_utils.cc#L61)):*
<details><summary><i>Previously, efoxepstein (Eli Fox-Epstein) wrote…</i></summary><blockquote>

For those less familiar with C++ (like me), maybe include a link in a comment to https://en.wikipedia.org/wiki/Erase%E2%80%93remove_idiom.
</blockquote></details>

Done.
___
*[src/test/cc/wfa/virtual_people/training/model_compiler/compiler_test.cc, line 425 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/9#-Mn8ErIOA4DcnIbfVcf6:-MnXQlUJDisM7ASKSdhp:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/2c0a17121e2462237bc6f89ddbe4d973c268d265/src/test/cc/wfa/virtual_people/training/model_compiler/compiler_test.cc#L425)):*
<details><summary><i>Previously, efoxepstein (Eli Fox-Epstein) wrote…</i></summary><blockquote>

These textprotos are pretty large. Consider moving them into `.textproto` files? That can be useful because we get syntax highlighting, validation, etc. in IDEs for textprotos.
</blockquote></details>

Done.
___
*[src/test/cc/wfa/virtual_people/training/model_compiler/test_data/activity_density_function.textproto, line 5 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/9#-Mn8F3hhD3P-PdC9hckL:-MnXQm8k5YRR8lsaXHoO:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/2c0a17121e2462237bc6f89ddbe4d973c268d265/src/test/cc/wfa/virtual_people/training/model_compiler/test_data/activity_density_function.textproto#L5)):*
<details><summary><i>Previously, efoxepstein (Eli Fox-Epstein) wrote…</i></summary><blockquote>

`identifier` looks like it's spelled wrong
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

tcsnfkx committed:
Format BUILD.bazel and compile.cc.

tcsnfkx committed:
Fix the return in CompileAdf.

efoxepstein reviewed (approved):
> Reviewed 16 of 16 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/9)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @EvgSkv, @kunyuchen, @lenjoy, @mpatel24, @my2457, and @wliue)



<!-- Sent from Reviewable.io -->

wliue reviewed (changes requested):
> Reviewed 6 of 16 files at r1, 16 of 16 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/9)* status: all files reviewed, 6 unresolved discussions (waiting on @EvgSkv, @kunyuchen, @lenjoy, @mpatel24, @my2457, and @tcsnfkx)

___
*[src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc, line 258 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/9#-MncBj-f-bHhxNDZ-R-1:-MncBj-f-bHhxNDZ-R-2:bdkjysi) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/756984ae714f22bf393bd45ff7b6a53e09b87c57/src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc#L258)):*
> ```
> NormalizeIfInError
> ```

How about NormalizeIfSumCloseToOne? The current name sounds like it will normalize if there is an error.
___
*[src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc, line 277 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/9#-Mnc4bJT4T1Ozg3FnmYV:-Mnc4bJT4T1Ozg3FnmYW:b9ki2e7) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/756984ae714f22bf393bd45ff7b6a53e09b87c57/src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc#L277)):*
<details><summary><i alt="summary">Quoted 10 lines of code…</i></summary>

> ```
>   std::vector<double> boundaries(alphas.size() + 1, 0.0);
>   for (int i = 0; i < alphas.size(); ++i) {
>     boundaries[i + 1] = population_sum * alphas[i] + boundaries[i];
>   }
> 
>   std::vector<uint64_t> discretized_boundaries(boundaries.size(), 0);
>   for (int i = 0; i < boundaries.size(); ++i) {
>     discretized_boundaries[i] = static_cast<uint64_t>(
>         round(boundaries[i] / discretization) * discretization);
>   }
> ```
</details>


Just want to point out that this doesn't necessarily minimize the error.
For example, 
population_sum = 3000, alphas = [0.24, 0.25, 0.51], discretization = 1000
will split into [1000, 0, 2000].

Maybe add a TODO to improve.
___
*[src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc, line 298 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/9#-Mnc7luWEU2DiLOzpz6R:-Mnc7luWEU2DiLOzpz6S:b-c6mfy) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/756984ae714f22bf393bd45ff7b6a53e09b87c57/src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc#L298)):*
> ```cpp
> SplitRecordsByDeltaPools(const std::vector<uint64_t>& delta_pool_sizes,
>                          const std::vector<CensusRecord*>& records) {
>   int next_record_index = 0;
> ```

Nit: Could add a high-level description.
___
*[src/main/cc/wfa/virtual_people/training/model_compiler/field_filter_utils.h, line 121 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/9#-MncH2qr0XOLjJsfU2j-:-MncH2qr0XOLjJsfU2j0:b-gubev2) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/756984ae714f22bf393bd45ff7b6a53e09b87c57/src/main/cc/wfa/virtual_people/training/model_compiler/field_filter_utils.h#L121)):*
<details><summary><i alt="summary">Quoted 16 lines of code…</i></summary>

> ```
> // Example 3:
> // If filter is
> //   op: AND
> //   sub_filters {
> //     op: EQUAL
> //     name: "a"
> //     value: "1"
> //   }
> //   sub_filters {
> //     op: EQUAL
> //     name: "a"
> //     value: "2"
> //   }
> // And name is "a"
> // The filter will be
> //   op: TRUE
> ```
</details>

Something is wrong with this example: it would never pass in original form, but will be converted to TRUE filter.
This may not be a problem in practice.
___
*[src/main/cc/wfa/virtual_people/training/model_compiler/field_filter_utils.cc, line 42 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/9#-MncItpPDc9wukKmzap8:-MncItpPDc9wukKmzap9:b-c7iymg) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/756984ae714f22bf393bd45ff7b6a53e09b87c57/src/main/cc/wfa/virtual_people/training/model_compiler/field_filter_utils.cc#L42)):*
> ```cpp
>       absl::StatusOr<std::string> value =
>           GetValueOfEqualFilter(sub_filter, name);
>       if (value.ok()) {
> ```

Potentially this could mask a config error that there are inconsistent EQUAL filters, as shown in the current example in .h file.
Maybe extract all matches and validate that all values are the same?
___
*[src/test/cc/wfa/virtual_people/training/model_compiler/compiler_test.cc, line 450 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/9#-MncLga8BJI-1sOAIPql:-MncLga8BJI-1sOAIPqm:b3cf8cx) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/756984ae714f22bf393bd45ff7b6a53e09b87c57/src/test/cc/wfa/virtual_people/training/model_compiler/compiler_test.cc#L450)):*
> ```cpp
>               StatusIs(absl::StatusCode::kInvalidArgument,
>                        "Neither verbatim nor from_file is set"));
> }
> ```

Could add some test cases for specific scenarios:
- discretization
- sum not close to 1
- readjust probability for empty pool
- empty pool for kappa < 1


<!-- Sent from Reviewable.io -->

tcsnfkx committed:
Fix based on comments.

Rename function NormalizeIfInError to NormalizeIfSumCloseToOne.
Update comments.
Add test cases for
  * discretization
  * sum not close to 1
  * redistribute probabilities for empty pools
  * empty pool for kappa < 1

tcsnfkx reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/9)* status: 17 of 29 files reviewed, 5 unresolved discussions (waiting on @efoxepstein, @EvgSkv, @kunyuchen, @lenjoy, @mpatel24, @my2457, and @wliue)

___
*[src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc, line 258 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/9#-MncBj-f-bHhxNDZ-R-1:-MndQxCU5s3xyRHSNdxd:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/756984ae714f22bf393bd45ff7b6a53e09b87c57/src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc#L258)):*
<details><summary><i>Previously, wliue (Wei Liu) wrote…</i></summary><blockquote>

> ```
> NormalizeIfInError
> ```

How about NormalizeIfSumCloseToOne? The current name sounds like it will normalize if there is an error.
</blockquote></details>

Done.
___
*[src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc, line 277 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/9#-Mnc4bJT4T1Ozg3FnmYV:-MndQy3E6lMA88BRsfXa:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/756984ae714f22bf393bd45ff7b6a53e09b87c57/src/main/cc/wfa/virtual_people/training/model_compiler/compiler.cc#L277)):*
<details><summary><i>Previously, wliue (Wei Liu) wrote…</i></summary><blockquote>

> ```
>   std::vector<double> boundaries(alphas.size() + 1, 0.0);
>   for (int i = 0; i < alphas.size(); ++i) {
>     boundaries[i + 1] = population_sum * alphas[i] + boundaries[i];
>   }
> 
>   std::vector<uint64_t> discretized_boundaries(boundaries.size(), 0);
>   for (int i = 0; i < boundaries.size(); ++i) {
>     discretized_boundaries[i] = static_cast<uint64_t>(
>         round(boundaries[i] / discretization) * discretization);
>   }
> ```

Just want to point out that this doesn't necessarily minimize the error.
For example, 
population_sum = 3000, alphas = [0.24, 0.25, 0.51], discretization = 1000
will split into [1000, 0, 2000].

Maybe add a TODO to improve.
</blockquote></details>

Done.
___
*[src/main/cc/wfa/virtual_people/training/model_compiler/field_filter_utils.h, line 121 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/9#-MncH2qr0XOLjJsfU2j-:-MndEjgW3pLye-e0GqPX:bkpp637) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/756984ae714f22bf393bd45ff7b6a53e09b87c57/src/main/cc/wfa/virtual_people/training/model_compiler/field_filter_utils.h#L121)):*
<details><summary><i>Previously, wliue (Wei Liu) wrote…</i></summary><blockquote>

> ```
> // Example 3:
> // If filter is
> //   op: AND
> //   sub_filters {
> //     op: EQUAL
> //     name: "a"
> //     value: "1"
> //   }
> //   sub_filters {
> //     op: EQUAL
> //     name: "a"
> //     value: "2"
> //   }
> // And name is "a"
> // The filter will be
> //   op: TRUE
> ```
Something is wrong with this example: it would never pass in original form, but will be converted to TRUE filter.
This may not be a problem in practice.
</blockquote></details>

I think it makes sense that the converted filter is a TRUE filter, because we are deleting some of the sub-filters.
___
*[src/main/cc/wfa/virtual_people/training/model_compiler/field_filter_utils.cc, line 42 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/9#-MncItpPDc9wukKmzap8:-MndFd5v3wMJrUT68zd0:btr1ki8) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/756984ae714f22bf393bd45ff7b6a53e09b87c57/src/main/cc/wfa/virtual_people/training/model_compiler/field_filter_utils.cc#L42)):*
<details><summary><i>Previously, wliue (Wei Liu) wrote…</i></summary><blockquote>

Potentially this could mask a config error that there are inconsistent EQUAL filters, as shown in the current example in .h file.
Maybe extract all matches and validate that all values are the same?
</blockquote></details>

Updated comments.
___
*[src/test/cc/wfa/virtual_people/training/model_compiler/compiler_test.cc, line 450 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/9#-MncLga8BJI-1sOAIPql:-MndQvOZ-ZnksypKs4gC:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/756984ae714f22bf393bd45ff7b6a53e09b87c57/src/test/cc/wfa/virtual_people/training/model_compiler/compiler_test.cc#L450)):*
<details><summary><i>Previously, wliue (Wei Liu) wrote…</i></summary><blockquote>

Could add some test cases for specific scenarios:
- discretization
- sum not close to 1
- readjust probability for empty pool
- empty pool for kappa < 1
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

tcsnfkx committed:
Apply clang-format.

tcsnfkx committed:
Format BUILD.bazel.

efoxepstein reviewed (changes requested):
> Reviewed 12 of 12 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/9)* status: all files reviewed, 8 unresolved discussions (waiting on @EvgSkv, @kunyuchen, @lenjoy, @mpatel24, @my2457, @tcsnfkx, and @wliue)

___
*[src/test/cc/wfa/virtual_people/training/model_compiler/compiler_test.cc, line 414 at r3](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/9#-MnfAjgo7MMW-CDnEr3t:-MnfAjgo7MMW-CDnEr3u:b-4resfl) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/86146a516ec8d8604419f0fe6789d215e48df2b2/src/test/cc/wfa/virtual_people/training/model_compiler/compiler_test.cc#L414)):*
> ```cpp
> TEST(CompileTest, PopulationNodeDiscretization) {
>   ModelNodeConfig config;
>   ASSERT_TRUE(
> ```

Can you replace `ASSERT_TRUE(foo.ok())` with `ASSERT_THAT(foo, IsOk())` in this entire repo? (Maybe in a follow-up PR?) This will lead to better error messages in the event that it's not OK.
___
*[src/test/cc/wfa/virtual_people/training/model_compiler/test_data/BUILD.bazel, line 1 at r3](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/9#-MnfBGpA2n5XAo5k2yKp:-MnfBGpA2n5XAo5k2yKq:bqe75z4) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/86146a516ec8d8604419f0fe6789d215e48df2b2/src/test/cc/wfa/virtual_people/training/model_compiler/test_data/BUILD.bazel#L1)):*
> ```Starlark
> exports_files([
> ```

Would it work to do`exports_files(glob["*.textproto"])`?
___
*[src/test/cc/wfa/virtual_people/training/model_compiler/test_data/compiled_node_for_population_node_redistribute_probabilities_for_empty_pools.textproto, line 2 at r3](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/9#-MnfC4jIE1xU5yDoCJka:-MnfC4jIE1xU5yDoCJkb:by7l7zt) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/86146a516ec8d8604419f0fe6789d215e48df2b2/src/test/cc/wfa/virtual_people/training/model_compiler/test_data/compiled_node_for_population_node_redistribute_probabilities_for_empty_pools.textproto#L2)):*
> ```
> name: "node1"
> branch_node {
> ```

What do you think about building a tool to spit out a Graphviz representation of the various trees? We have a [helper library](https://github.com/world-federation-of-advertisers/common-jvm/blob/main/src/main/kotlin/org/wfanet/measurement/common/graphviz/Dot.kt) for this.

It could be useful to visualize the branching structure, which I personally find hard to do in large textprotos.

This could definitely happen in a different PR.


<!-- Sent from Reviewable.io -->

wliue reviewed (changes requested):
> Reviewed 11 of 12 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/9)* status: all files reviewed, 9 unresolved discussions (waiting on @EvgSkv, @kunyuchen, @lenjoy, @mpatel24, @my2457, @tcsnfkx, and @wliue)

___
*[src/main/cc/wfa/virtual_people/training/model_compiler/field_filter_utils.h, line 121 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/9#-MncH2qr0XOLjJsfU2j-:-Mnfjon51EFA9diSbbGH:bvcm4jw) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/756984ae714f22bf393bd45ff7b6a53e09b87c57/src/main/cc/wfa/virtual_people/training/model_compiler/field_filter_utils.h#L121)):*
<details><summary><i>Previously, tcsnfkx wrote…</i></summary><blockquote>

I think it makes sense that the converted filter is a TRUE filter, because we are deleting some of the sub-filters.
</blockquote></details>

This behavior is caused by current implementation, it doesn't mean that it is the expected behavior. We should at least state this issue explicitly in the comment.
___
*[src/main/cc/wfa/virtual_people/training/model_compiler/field_filter_utils.h, line 33 at r3](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/9#-MnfkZly6GwGAlVCaBPw:-MnfkZly6GwGAlVCaBPx:b6y5b69) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/86146a516ec8d8604419f0fe6789d215e48df2b2/src/main/cc/wfa/virtual_people/training/model_compiler/field_filter_utils.h#L33)):*
> ```objc
> // If @filter.op is AND, apply the check to each of the @filter.sub_filters, and
> // return the value if any matching entity is found.
> // Only the value of the first matching is returned.
> ```

Also mention that this wouldn't catch config errors that there are multiple sub_filters on the same field with different values.
___
*[src/test/cc/wfa/virtual_people/training/model_compiler/compiler_test.cc, line 508 at r3](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/9#-MnflJYQ-o-A7-gAGwyN:-MnflJYQ-o-A7-gAGwyO:btdj6iq) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/86146a516ec8d8604419f0fe6789d215e48df2b2/src/test/cc/wfa/virtual_people/training/model_compiler/compiler_test.cc#L508)):*
> ```cpp
> }
> 
> TEST(CompileTest, PopulationNodeAphaNotSumToOne) {
> ```

Apha -> Alpha


<!-- Sent from Reviewable.io -->

tcsnfkx committed:
Fix based on comments.

Update comments.
Replace ASSERT_TRUE(foo.ok()) with ASSERT_THAT(foo, IsOk()).
Fix typo.
Use glob in BUILD.bazel.

tcsnfkx reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/9)* status: 26 of 29 files reviewed, 7 unresolved discussions (waiting on @efoxepstein, @EvgSkv, @kunyuchen, @lenjoy, @mpatel24, @my2457, and @wliue)

___
*[src/main/cc/wfa/virtual_people/training/model_compiler/field_filter_utils.h, line 121 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/9#-MncH2qr0XOLjJsfU2j-:-Mng4F2550LL1czL0wy8:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/756984ae714f22bf393bd45ff7b6a53e09b87c57/src/main/cc/wfa/virtual_people/training/model_compiler/field_filter_utils.h#L121)):*
<details><summary><i>Previously, wliue (Wei Liu) wrote…</i></summary><blockquote>

This behavior is caused by current implementation, it doesn't mean that it is the expected behavior. We should at least state this issue explicitly in the comment.
</blockquote></details>

Done.
___
*[src/main/cc/wfa/virtual_people/training/model_compiler/field_filter_utils.h, line 33 at r3](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/9#-MnfkZly6GwGAlVCaBPw:-Mng4D_u9HFji8Ta8-7I:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/86146a516ec8d8604419f0fe6789d215e48df2b2/src/main/cc/wfa/virtual_people/training/model_compiler/field_filter_utils.h#L33)):*
<details><summary><i>Previously, wliue (Wei Liu) wrote…</i></summary><blockquote>

Also mention that this wouldn't catch config errors that there are multiple sub_filters on the same field with different values.
</blockquote></details>

Done.
___
*[src/test/cc/wfa/virtual_people/training/model_compiler/compiler_test.cc, line 508 at r3](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/9#-MnflJYQ-o-A7-gAGwyN:-Mng4AX30yekm79gHRfE:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/86146a516ec8d8604419f0fe6789d215e48df2b2/src/test/cc/wfa/virtual_people/training/model_compiler/compiler_test.cc#L508)):*
<details><summary><i>Previously, wliue (Wei Liu) wrote…</i></summary><blockquote>

Apha -> Alpha
</blockquote></details>

Done.
___
*[src/test/cc/wfa/virtual_people/training/model_compiler/test_data/BUILD.bazel, line 1 at r3](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/9#-MnfBGpA2n5XAo5k2yKp:-Mng4BU_2sZI8lyFIpRj:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/86146a516ec8d8604419f0fe6789d215e48df2b2/src/test/cc/wfa/virtual_people/training/model_compiler/test_data/BUILD.bazel#L1)):*
<details><summary><i>Previously, efoxepstein (Eli Fox-Epstein) wrote…</i></summary><blockquote>

Would it work to do`exports_files(glob["*.textproto"])`?
</blockquote></details>

Done.
___
*[src/test/cc/wfa/virtual_people/training/model_compiler/test_data/compiled_node_for_population_node_redistribute_probabilities_for_empty_pools.textproto, line 2 at r3](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/9#-MnfC4jIE1xU5yDoCJka:-Mng4H-01hkitDQAA29i:b-l76rls) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/86146a516ec8d8604419f0fe6789d215e48df2b2/src/test/cc/wfa/virtual_people/training/model_compiler/test_data/compiled_node_for_population_node_redistribute_probabilities_for_empty_pools.textproto#L2)):*
<details><summary><i>Previously, efoxepstein (Eli Fox-Epstein) wrote…</i></summary><blockquote>

What do you think about building a tool to spit out a Graphviz representation of the various trees? We have a [helper library](https://github.com/world-federation-of-advertisers/common-jvm/blob/main/src/main/kotlin/org/wfanet/measurement/common/graphviz/Dot.kt) for this.

It could be useful to visualize the branching structure, which I personally find hard to do in large textprotos.

This could definitely happen in a different PR.
</blockquote></details>

That makes sense. How about an independent task to add such tool in common-cpp?


<!-- Sent from Reviewable.io -->

efoxepstein reviewed (approved):
> Reviewed 3 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/9)* status: all files reviewed, 6 unresolved discussions (waiting on @EvgSkv, @kunyuchen, @lenjoy, @mpatel24, @my2457, and @wliue)

___
*[src/test/cc/wfa/virtual_people/training/model_compiler/test_data/compiled_node_for_population_node_redistribute_probabilities_for_empty_pools.textproto, line 2 at r3](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/9#-MnfC4jIE1xU5yDoCJka:-MngEBB0E1bRYzTepIqg:b-hfghm4) ([raw file](https://github.com/world-federation-of-advertisers/virtual-people-training/blob/86146a516ec8d8604419f0fe6789d215e48df2b2/src/test/cc/wfa/virtual_people/training/model_compiler/test_data/compiled_node_for_population_node_redistribute_probabilities_for_empty_pools.textproto#L2)):*
<details><summary><i>Previously, tcsnfkx wrote…</i></summary><blockquote>

That makes sense. How about an independent task to add such tool in common-cpp?
</blockquote></details>

Sounds good


<!-- Sent from Reviewable.io -->

wliue reviewed (approved):
> Reviewed 1 of 12 files at r3, 3 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/9)* status: all files reviewed, 2 unresolved discussions (waiting on @EvgSkv, @kunyuchen, @lenjoy, @mpatel24, @my2457, and @wliue)



<!-- Sent from Reviewable.io -->

tcsnfkx reviewed (commented):
> Dismissed @wliue from 2 discussions.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/virtual-people-training/9)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @EvgSkv, @kunyuchen, @lenjoy, and @my2457)



<!-- Sent from Reviewable.io -->
